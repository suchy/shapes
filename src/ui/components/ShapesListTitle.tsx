import styled from 'styled-components';

import { ShapeColors, ShapeShapes } from '../../domain/shapes/shape';

import { useShallowSelector } from '../hooks/use-shallow-selector';

import { filtersSelector } from '../state/selectors';

const Title = styled.h2`
  text-align: center;
  margin-bottom: 4rem;
  font-size: 3.5rem;
  font-weight: bold;
  font-style: italic;
  text-transform: uppercase;

  @media (min-width: 1024px) {
    text-align: left;
    margin-bottom: 8rem;
  }
`;

export const ShapesListTitle = () => {
  const { colorFilters, shapeFilters } = useShallowSelector(filtersSelector);

  const allColorsCount = Object.keys(ShapeColors).length;
  const selectedColorsCount = colorFilters.length;

  const allShapesCount = Object.keys(ShapeShapes).length;
  const selectedShapesCount = shapeFilters.length;

  const color = colorFilters[0];
  const shape = shapeFilters[0];

  let title = `${shape} ${color} items:`;

  if (
    selectedColorsCount === allColorsCount &&
    selectedShapesCount === allShapesCount
  ) {
    title = 'All items:';
  }

  if (
    selectedColorsCount === allColorsCount &&
    selectedShapesCount < allShapesCount &&
    selectedShapesCount > 1
  ) {
    title = 'Multiple items:';
  }

  if (
    selectedShapesCount === allShapesCount &&
    selectedColorsCount < allColorsCount &&
    selectedColorsCount > 1
  ) {
    title = 'Multiple items:';
  }

  if (selectedShapesCount === allShapesCount && selectedColorsCount === 1) {
    const color = colorFilters[0];
    title = `All ${color} items:`;
  }

  if (selectedColorsCount === allColorsCount && selectedShapesCount === 1) {
    const shape = shapeFilters[0];
    title = `All ${shape} items:`;
  }

  if (
    selectedShapesCount < allShapesCount &&
    selectedShapesCount > 1 &&
    selectedColorsCount === 1
  ) {
    const color = colorFilters[0];
    title = `Multiple ${color} items:`;
  }

  if (
    selectedColorsCount < allColorsCount &&
    selectedColorsCount > 1 &&
    selectedShapesCount === 1
  ) {
    const shape = shapeFilters[0];
    title = `Multiple ${shape} items:`;
  }

  return <Title>{title}</Title>;
};
