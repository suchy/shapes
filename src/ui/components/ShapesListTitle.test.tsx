import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReducer, configureStore } from '@reduxjs/toolkit';

import { ShapesListTitle } from './ShapesListTitle';

jest.mock('../../domain/shapes/shape', () => ({
  ShapeColors: { red: 'red', purple: 'purple', blue: 'blue' },
  ShapeShapes: { round: 'round', square: 'square', triangle: 'triangle' }
}));

describe('ShapesListTitle', () => {
  it('should match snapshot', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'purple', 'blue'],
          shapeFilters: ['round', 'square', 'triangle']
        },
        jest.fn()
      )
    });

    const { container } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show title if there is one color and one shape', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red'],
          shapeFilters: ['round']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('round red items:')).toBeVisible();
  });

  it('should show title if there are all colors and shapes', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'purple', 'blue'],
          shapeFilters: ['round', 'square', 'triangle']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('All items:')).toBeVisible();
  });

  it('should show title if there are all colors and more than one shape', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'purple', 'blue'],
          shapeFilters: ['round', 'square']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('Multiple items:')).toBeVisible();
  });

  it('should show title if there are all shapes and more than one color', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'purple'],
          shapeFilters: ['round', 'square', 'triangle']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('Multiple items:')).toBeVisible();
  });

  it('should show title if there are all shapes and one color', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red'],
          shapeFilters: ['round', 'square', 'triangle']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('All red items:')).toBeVisible();
  });

  it('should show title if there are all colors and one shape', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'purple', 'blue'],
          shapeFilters: ['round']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('All round items:')).toBeVisible();
  });

  it('should show title if there is more than one shape and one color', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red'],
          shapeFilters: ['round', 'square']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('Multiple red items:')).toBeVisible();
  });

  it('should show title if there is more than one color and one shape', () => {
    const store = configureStore({
      reducer: createReducer(
        {
          colorFilters: ['red', 'blue'],
          shapeFilters: ['round']
        },
        jest.fn()
      )
    });

    const { getByText } = render(
      <Provider store={store}>
        <ShapesListTitle />
      </Provider>
    );

    expect(getByText('Multiple round items:')).toBeVisible();
  });
});
