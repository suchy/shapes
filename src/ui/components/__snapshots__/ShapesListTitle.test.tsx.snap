// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapesListTitle should match snapshot 1`] = `
<h2
  class="sc-bdfBwQ dHzUuQ"
>
  All items:
</h2>
`;
